File: topics/src/main/kotlin/dev/kortex/myinfo/topics/domain/usecase/WebAddress.kt
package dev.kortex.myinfo.topics.domain.usecase

import java.net.URI

/** Recognises web addresses in pasted text, with or without a scheme. */
internal object WebAddress {
    // host.tld with optional port and path; the scheme is added before parsing.
    private val schemeless = Regex(
        """^([a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z]{2,}(:\d{1,5})?([/?#]\S*)?$""",
        RegexOption.IGNORE_CASE,
    )

    /** [text] as an http(s) address, `https://` added when it has no scheme; null when it isn't one. */
    fun parse(text: String): URI? {
        val trimmed = text.trim()
        if (trimmed.isEmpty() || trimmed.any(Char::isWhitespace)) return null
        val candidate = when {
            trimmed.startsWith("http://", ignoreCase = true) || trimmed.startsWith("https://", ignoreCase = true) -> trimmed
            schemeless.matches(trimmed) -> "https://$trimmed"
            else -> return null
        }
        val uri = runCatching { URI(candidate) }.getOrNull() ?: return null
        return uri.takeUnless { it.host.isNullOrEmpty() }
    }
}

File: functions/src/webAddress.ts
/**
 * Port of `WebAddress.parse` in topics/…/WebAddress.kt: [text] as an http(s) address, with
 * `https://` added when it has no scheme, or null when it isn't a web address.
 */
export function parseWebAddress(text: string): string | null {
  const trimmed = text.trim();
  if (trimmed.length === 0 || /\s/.test(trimmed)) return null;
  let candidate: string;
  if (/^https?:\/\//i.test(trimmed)) {
    candidate = trimmed;
  } else if (SCHEMELESS.test(trimmed)) {
    candidate = `https://${trimmed}`;
  } else {
    return null;
  }
  try {
    return new URL(candidate).hostname.length > 0 ? candidate : null;
  } catch {
    return null;
  }
}

// host.tld with optional port and path; the scheme is added before parsing.
const SCHEMELESS = /^([a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z]{2,}(:\d{1,5})?([/?#]\S*)?$/i;
